contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,475 @@
+
+## Use Cases
+
+  * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.)
+    in my pod.
+  * I should be able to use that device without writing custom Kubernetes code.
+  * I want a consistent and portable solution to consuming hardware devices
+    across k8s clusters
+
+## Objectives
+
+1. Add support for vendor specific Devices in kubelet:
+    * Through a pluggable mechanism
+    * Which allows discovery and monitoring of devices
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API
+3. Define a versioning mechanism oor this new API
+t
+
+## Non Objectives
+1. Advanced scheduling and resource selection (solved through [#782](https://github.com/kubernetes/community/pull/782))
+   We will only try to give basic selection primitives to the devices
+2. Metrics this should be the job of cadvisor and should probably either be
+   addressed there (cadvisor) or if people feel there is a case to be made
+   for it being addressed in the Device Plugin in a follow up proposal.
+
+## Stories
+
+### Vendor story
+
+Kubernetes provides to vendors a mechanism called device plugins to:
+  * advertise devices
+  * monitor devices (currently perform health checks)
+  * hook into the runtime to instruct Kubelet what are the steps to
+    take in order to make the device available (or cleanup the device)
+
+A device plugin at it's core is a simple gRPC server usually running in
+a container and deployed across clusters through a daemonSet.
+
+```gRPC
+service DevicePlugin {
+	rpc Init(Empty) returns (Empty) {}
+
+	rpc Discover(Empty) returns (stream Device) {}
+	rpc Monitor(Empty) returns (stream DeviceHealth) {}
+
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+	rpc Deallocate(DeallocateRequest) returns (Empty) {}
+}
+
+```
+
+The gRPC server that the device plugin must implement is expected to
+be advertised on a unix socket in a mounted hostPath (e.g:
+`/var/run/kubernetes/vendor.sock`).
+
+Finally, to notify Kubelet of the existence of the device plugin,
+the vendor's device plugin will have to make a request to Kubelet's
+onwn gRPC server.
+Only then will kubelet start interacting with the vendor's device plugin
+through the gRPC apis.
+
+### End User story
+
+When setting up the cluster the admin knows what kind of devices are present
+on the different machines and therefore can select what devices they want to
+enable.
+
+The cluster admins knows his cluster has Nvidia GPUs therefore he deploys
+the nvidia device plugin through:
+`kubectl create -f nvidia.io/device-plugin.yml`
+
+The device plugin lands on all the nodes of the cluster and if it detects that
+there are no GPUs it terminates. However, when there are GPUs it reports them
+to Kubelet.
+For device plugins reporting non-GPU Devices these are advertised as
+OIRs and selected through the same method.
```

Comment:
> Can non-gpu devices be advertised using the above DaemonSet mechanism? 

I don't understand your question about DaemonSet ? DaemonSet is just a way to deploy a Device Plugin. The plugin then advertise the resource to Kubelet.

Currently the idea is to have GPUs advertised by the Nvidia device plugin to be advertised by the node as "alpha.kubernetes.io/nvidia-gpu" and all other devices as "pod.alpha.kubernetes.io/opaque-int-resource-my-device".

We are thinking of actually changing the name to "extensions.kubernetes.io/my-device" but this is really another discussion :)